docs/user-guide/materialization.md

```diff
@@ -0,0 +1,236 @@
+  until it has a list of jobs and the list of jobs cannot depends
+  on the nix being present (although this is often blamed on IFD
+  it would be the same wrote out JSON files and read them in)
+
+## When is it ok to materialize?
+
+* The nix is unlikely to change frequently (and when it does you
+  are happy to manually update it).
+
+* You are happy to script something to update the materialized
+  nix files automatically.
+
+## How can we materialize the nix files?
+
+Lets say we want to build hlint.  We might start with an hlint
+file that looks like this:
+
+```nix
+((import ./nixpkgs (import ./.)).haskell-nix.hackage-package {
+  name = "hlint";
+  version = "2.2.4";
+}).components.exes.hlint
+```
+
+Building this may result in a lot of output, but if youb build
+it again it should give just:
+
+```
+$ nix-build hlint.nix 
+trace: Using latest index state for hlint!
+trace: Using index-state: 2019-12-09T00:00:00Z for hlint
+/nix/store/7wwqq5v42gm6iiz2d3ngbnkyw7i4py11-hlint-2.2.4-exe-hlint
+```
+
+To materialize the nix files we need to take care to pin down the
+inputs.  For `cabalProject` and `hackage-package` this means
+we must specify the `index-state` of hackage we want to use:
+
+```nix
+((import ./nixpkgs (import ./.)).haskell-nix.hackage-package {
+  name = "hlint";
+  version = "2.2.4";
+  index-state = "2019-12-03T00:00:00Z";
+}).components.exes.hlint
+```
+
+Now if we build again we get a hint telling use how to
+calculate a suitable sha256 hash to turn the derivation
+containing the nix files into a fixed output derivation:
+
+```
+$ nix-build hlint.nix 
+trace: Using index-state: 2019-12-03T00:00:00Z for hlint
+trace: Get `plan-sha256` with `nix-hash --base32 --type sha256 /nix/store/qk1fvza1alkvs51vzmpjp2xsg8xklyxk-hlint-plan-to-nix-pkgs/`
+/nix/store/7wwqq5v42gm6iiz2d3ngbnkyw7i4py11-hlint-2.2.4-exe-hlint
+$ nix-hash --base32 --type sha256 /nix/store/qk1fvza1alkvs51vzmpjp2xsg8xklyxk-hlint-plan-to-nix-pkgs/
+1a4rhv3h2daz6dzwzfl3w7l1v556n7aqfiagw6m0rvqq230iabss
+```
+
+We can add the hash as `plan-sha256` or (`stack-sha256` for
+`stackProject`)
+
+```nix
+((import ./nixpkgs (import ./.)).haskell-nix.hackage-package {
+  name = "hlint";
+  version = "2.2.4";
+  index-state = "2019-12-03T00:00:00Z";
+  plan-sha256 = "1a4rhv3h2daz6dzwzfl3w7l1v556n7aqfiagw6m0rvqq230iabss";
+}).components.exes.hlint
+```
+
+Just adding the hash might help reuse of the cached nix, but nix will
+still calculate all the dependencies (which can add seconds to
+`nix-build` and `nix-shell` commands when no ohter work is needed)
+and users who do not yet have the dependencies in their store will have
+to wait while they are built or downloaded.
+
+Running nix build again gives us a hint on what we can do next:
+
+```
+$ nix-build hlint.nix 
+trace: Using index-state: 2019-12-03T00:00:00Z for hlint
+trace: To materialize copy /nix/store/0xalcphb7ifvy5fc9dpwj40fij6nn5av-hlint-plan-to-nix-pkgs
```

Comment:
I think we need some wording adjustment here
copy ... to ...

Reply:
I had changed that in the code already, just forgot to update the docs.   Updated now.